File: implementation/scipamato/src/test/java/ch/difty/scipamato/entity/CodeTest.java
package ch.difty.scipamato.entity;

import static org.assertj.core.api.Assertions.assertThat;

import java.time.LocalDateTime;

import javax.validation.ConstraintViolation;

import org.junit.Test;

public class CodeTest extends Jsr303ValidatedEntityTest<Code> {

    private static final String CODE1 = "code1";
    private static final String JAVAX_VALIDATION_CONSTRAINTS_NOT_NULL_MESSAGE = "{javax.validation.constraints.NotNull.message}";
    private static final String CODECLASS10 = "codeclass10";
    private static final LocalDateTime CREATED = LocalDateTime.parse("2017-01-01T08:01:33.821");
    private static final LocalDateTime LAST_MOD = LocalDateTime.parse("2017-02-02T08:01:33.821");

    private final Code c = new Code("1A", CODE1, null, false, 1, "c1", "", 1);

    private void verifySuccessfulValidation(Code code) {
        validate(code);
        assertThat(getViolations()).isEmpty();
    }

    private void validateAndAssertFailure(Code code, final String field, final Object invalidValue, final String msg) {
        validate(code);

        assertThat(getViolations()).isNotEmpty().hasSize(1);
        ConstraintViolation<Code> violation = getViolations().iterator().next();
        assertThat(violation.getMessageTemplate()).isEqualTo(msg);
        assertThat(violation.getInvalidValue()).isEqualTo(invalidValue);
        assertThat(violation.getPropertyPath().toString()).isEqualTo(field);
    }

    @Test
    public void constructing_withAllValues_populatesCodeClass() {
        Code c1 = new Code("C1", "c1", null, false, 10, "cc10", CODECLASS10, 2);
        assertThat(c1.getCodeClass()).isNotNull();
        assertThat(c1.getCodeClass().getId()).isEqualTo(10);
        assertThat(c1.getCodeClass().getName()).isEqualTo("cc10");
        assertThat(c1.getCodeClass().getDescription()).isEqualTo(CODECLASS10);
    }

    @Test
    public void constructing_withNullCodeClassId_leaveCodeClassNull() {
        Code c1 = new Code("C1", "c1", null, false, null, "cc10", CODECLASS10, 1);
        assertThat(c1.getCodeClass()).isNull();
    }

    @Test
    public void validatingCode_beginValid_succeeds() {
        verifySuccessfulValidation(c);
    }

    @Test
    public void validatingCode_withNullCode_fails() {
        Code c1 = new Code(null, CODE1, null, false, 1, "c1", "", 1);
        validateAndAssertFailure(c1, Code.CODE, null, JAVAX_VALIDATION_CONSTRAINTS_NOT_NULL_MESSAGE);
    }

    @Test
    public void validatingCode_withNullName_fails() {
        Code c1 = new Code("1A", null, null, false, 1, "c1", "", 1);
        validateAndAssertFailure(c1, Code.NAME, null, JAVAX_VALIDATION_CONSTRAINTS_NOT_NULL_MESSAGE);
    }

    @Test
    public void validatingCode_withNullCodeClass_fails() {
        Code c1 = new Code("1A", CODE1, null, false, null, null, null, 1);
        validateAndAssertFailure(c1, Code.CODE_CLASS, null, JAVAX_VALIDATION_CONSTRAINTS_NOT_NULL_MESSAGE);
    }

    @Test
    public void validatingCode_withWrongCodeFormat_fails() {
        Code c1 = new Code("xyz", CODE1, null, false, 1, "c1", "", 1);
        validateAndAssertFailure(c1, Code.CODE, "xyz", "{code.invalidCode}");
    }

    @Test
    public void testingToString() {
        final Code c1 = new Code("1A", CODE1, null, false, 1, "c1", "", 1, CREATED, 10, LAST_MOD, 20, 3);
        assertThat(c1.toString()).isEqualTo(
                "Code[code=1A,name=code1,comment=<null>,internal=false,codeClass=CodeClass[id=1],sort=1,created=2017-01-01T08:01:33.821,createdBy=10,lastModified=2017-02-02T08:01:33.821,lastModifiedBy=20,version=3]");
    }

    @Test
    public void cloning_copiesValues() {
        Code c1 = new Code("1A", CODE1, "foo", true, 1, "c1", "", 2);
        Code c2 = new Code(c1);
        assertThat(c2.getCode()).isEqualTo("1A");
        assertThat(c2.getName()).isEqualTo(CODE1);
        assertThat(c2.getComment()).isEqualTo("foo");
        assertThat(c2.isInternal()).isTrue();
        assertThat(c2.getCodeClass()).isEqualToComparingFieldByField(c1.getCodeClass());
        assertThat(c2.getSort()).isEqualTo(2);
    }

    @Test
    public void sameValues_makeEquality() {
        Code c2 = new Code(c);
        assertThat(c.equals(c2)).isTrue();
        assertThat(c2.equals(c)).isTrue();
        assertThat(c.hashCode()).isEqualTo(c2.hashCode());
    }

    @Test
    public void differingValues_makeInequality() {
        Code c1 = new Code("1A", CODE1, null, false, 1, "c1", "", 1);
        Code c2 = new Code("1B", CODE1, null, false, 1, "c1", "", 1);
        Code c3 = new Code("1A", "code2", null, false, 1, "c1", "", 1);
        Code c4 = new Code("1A", CODE1, null, false, 2, "c2", "", 1);
        Code c5 = new Code("1A", CODE1, null, false, 1, "c2", "", 2);
        Code c6 = new Code("1A", CODE1, "foo", false, 1, "c1", "", 1);
        Code c7 = new Code("1A", CODE1, null, true, 1, "c1", "", 1);

        assertThat(c1.equals(c2)).isFalse();
        assertThat(c1.equals(c3)).isFalse();
        assertThat(c1.equals(c4)).isFalse();
        assertThat(c1.equals(c5)).isFalse();
        assertThat(c1.equals(c6)).isFalse();
        assertThat(c1.equals(c7)).isFalse();
        assertThat(c2.equals(c3)).isFalse();
        assertThat(c2.equals(c4)).isFalse();
        assertThat(c2.equals(c5)).isFalse();
        assertThat(c2.equals(c6)).isFalse();
        assertThat(c2.equals(c7)).isFalse();
        assertThat(c3.equals(c4)).isFalse();
        assertThat(c3.equals(c5)).isFalse();
        assertThat(c3.equals(c6)).isFalse();
        assertThat(c3.equals(c7)).isFalse();
        assertThat(c4.equals(c5)).isFalse();
        assertThat(c4.equals(c6)).isFalse();
        assertThat(c4.equals(c7)).isFalse();
        assertThat(c5.equals(c6)).isFalse();
        assertThat(c5.equals(c7)).isFalse();
        assertThat(c6.equals(c7)).isFalse();

        assertThat(c1.hashCode()).isNotEqualTo(c2.hashCode());
        assertThat(c1.hashCode()).isNotEqualTo(c3.hashCode());
        assertThat(c1.hashCode()).isNotEqualTo(c4.hashCode());
        assertThat(c1.hashCode()).isNotEqualTo(c5.hashCode());
        assertThat(c1.hashCode()).isNotEqualTo(c6.hashCode());
        assertThat(c1.hashCode()).isNotEqualTo(c7.hashCode());
        assertThat(c2.hashCode()).isNotEqualTo(c3.hashCode());
        assertThat(c2.hashCode()).isNotEqualTo(c4.hashCode());
        assertThat(c2.hashCode()).isNotEqualTo(c5.hashCode());
        assertThat(c2.hashCode()).isNotEqualTo(c6.hashCode());
        assertThat(c2.hashCode()).isNotEqualTo(c7.hashCode());
        assertThat(c3.hashCode()).isNotEqualTo(c4.hashCode());
        assertThat(c3.hashCode()).isNotEqualTo(c5.hashCode());
        assertThat(c3.hashCode()).isNotEqualTo(c6.hashCode());
        assertThat(c3.hashCode()).isNotEqualTo(c7.hashCode());
        assertThat(c4.hashCode()).isNotEqualTo(c5.hashCode());
        assertThat(c4.hashCode()).isNotEqualTo(c6.hashCode());
        assertThat(c4.hashCode()).isNotEqualTo(c7.hashCode());
        assertThat(c5.hashCode()).isNotEqualTo(c6.hashCode());
        assertThat(c5.hashCode()).isNotEqualTo(c7.hashCode());
        assertThat(c6.hashCode()).isNotEqualTo(c7.hashCode());
    }

    @Test
    public void equalingToSpecialCases() {
        assertThat(c.equals(c)).isTrue();
        assertThat(c.equals(null)).isFalse();
        assertThat(c.equals("")).isFalse();
    }

    @Test
    public void displayValue() {
        assertThat(c.getDisplayValue()).isEqualTo("code1 (1A)");
    }

    @Test
    public void differingValues_withCodeNullOnOne() {
        Code c1 = new Code(null, CODE1, null, false, 1, "c1", "", 1);
        Code c2 = new Code("1A", CODE1, null, false, 1, "c1", "", 1);
        assertInequality(c1, c2);
    }

    private void assertInequality(Code c1, Code c2) {
        assertThat(c1.equals(c2)).isFalse();
        assertThat(c2.equals(c1)).isFalse();
        assertThat(c1.hashCode()).isNotEqualTo(c2.hashCode());
    }

    @Test
    public void differingValues_withCodeNullOnBoth() {
        Code c1 = new Code(null, CODE1, null, false, 1, "c1", "", 1);
        Code c2 = new Code(null, CODE1, null, false, 1, "c1", "", 1);
        assertEquality(c1, c2);
    }

    private void assertEquality(Code c1, Code c2) {
        assertThat(c1.equals(c2)).isTrue();
        assertThat(c2.equals(c1)).isTrue();
        assertThat(c1.hashCode()).isEqualTo(c2.hashCode());
    }

    @Test
    public void differingValues_withCodeClassNullOnOne() {
        Code c1 = new Code("1A", CODE1, null, false, null, "c1", "", 1);
        Code c2 = new Code("1A", CODE1, null, false, 1, "c1", "", 1);
        assertInequality(c1, c2);
    }

    @Test
    public void differingValues_withCodeClassNullOnBoth() {
        Code c1 = new Code("1A", CODE1, null, false, null, "c1", "", 1);
        Code c2 = new Code("1A", CODE1, null, false, null, "c1", "", 1);
        assertEquality(c1, c2);
    }

    @Test
    public void differingValues_withNameNullOnOne() {
        Code c1 = new Code("1A", null, null, false, 1, "c1", "", 1);
        Code c2 = new Code("1A", CODE1, null, false, 1, "c1", "", 1);
        assertInequality(c1, c2);
    }

    @Test
    public void differingValues_withNameNullOnBoth() {
        Code c1 = new Code("1A", null, null, false, 1, "c1", "", 1);
        Code c2 = new Code("1A", null, null, false, 1, "c1", "", 1);
        assertEquality(c1, c2);
    }

    @Test
    public void differingValues_withSameComment() {
        Code c1 = new Code("1A", CODE1, "foo", false, 1, "c1", "", 1);
        Code c2 = new Code("1A", CODE1, "foo", false, 1, "c1", "", 1);
        assertEquality(c1, c2);
    }

    @Test
    public void differingValues_withDifferingComment() {
        Code c1 = new Code("1A", CODE1, "foo", false, 1, "c1", "", 1);
        Code c2 = new Code("1A", CODE1, "bar", false, 1, "c1", "", 1);
        assertInequality(c1, c2);
    }

    @Test
    public void differingValues_withDifferingSort() {
        Code c1 = new Code("1A", CODE1, "", false, 1, "c1", "", 1);
        Code c2 = new Code("1A", CODE1, "", false, 1, "c1", "", 2);
        assertInequality(c1, c2);
    }
}

File: core/core-entity/src/test/kotlin/ch/difty/scipamato/core/entity/CodeTest.kt
package ch.difty.scipamato.core.entity

import ch.difty.scipamato.core.entity.Code.CodeFields.CODE
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeFalse
import org.amshove.kluent.shouldBeTrue
import org.amshove.kluent.shouldNotBeEqualTo
import org.amshove.kluent.shouldNotBeNull
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

internal class CodeTest : Jsr303ValidatedEntityTest<Code>(Code::class.java) {

    override fun newValidEntity(): Code {
        return Code("1A", CODE1, null, false, 1, "c1", "", 1, CREATED, 10, LAST_MOD, 20, 3)
    }

    override val toString =
        """Code[code=1A,name=code1,comment=<null>,internal=false,codeClass=CodeClass[id=1],sort=1,
            |createdBy=10,lastModifiedBy=20,created=2017-01-01T08:01:33.821,
            |lastModified=2017-02-02T08:01:33.821,version=3]""".trimMargin()
    override val displayValue = "code1 (1A)"

    @Test
    fun constructing_withAllValues_populatesCodeClass() {
        val c1 = Code("C1", "c1", null, false, 10, "cc10", CODECLASS10, 2)
        c1.codeClass.shouldNotBeNull()
        c1.codeClass.id shouldBeEqualTo 10
        c1.codeClass.name shouldBeEqualTo "cc10"
        c1.codeClass.description shouldBeEqualTo CODECLASS10
    }

    @Test
    fun validatingCode_withWrongCodeFormat_fails() {
        val c1 = Code("xyz", CODE1, null, false, 1, "c1", "", 1)
        validateAndAssertFailure(c1, CODE, "xyz", "{code.invalidCode}")
    }

    @Test
    fun cloning_copiesValues() {
        val c1 = Code("1A", CODE1, "foo", true, 1, "c1", "", 2)
        val c2 = Code(c1)
        c2.code shouldBeEqualTo "1A"
        c2.name shouldBeEqualTo CODE1
        c2.comment shouldBeEqualTo "foo"
        c2.isInternal.shouldBeTrue()
        c2.codeClass shouldBeEqualTo c1.codeClass
        c2.sort shouldBeEqualTo 2
    }

    @Test
    fun sameValues_makeEquality() {
        val c = newValidEntity()
        val c2 = Code(c)
        (c == c2).shouldBeTrue()
        (c2 == c).shouldBeTrue()
        c.hashCode() shouldBeEqualTo c2.hashCode()
    }

    @Test
    fun differingValues_makeInequality() {
        val c1 = Code("1A", CODE1, null, false, 1, "c1", "", 1)
        val c2 = Code("1B", CODE1, null, false, 1, "c1", "", 1)
        val c3 = Code("1A", "code2", null, false, 1, "c1", "", 1)
        val c4 = Code("1A", CODE1, null, false, 2, "c2", "", 1)
        val c5 = Code("1A", CODE1, null, false, 1, "c2", "", 2)
        val c6 = Code("1A", CODE1, "foo", false, 1, "c1", "", 1)
        val c7 = Code("1A", CODE1, null, true, 1, "c1", "", 1)

        (c1 == c2).shouldBeFalse()
        (c1 == c3).shouldBeFalse()
        (c1 == c4).shouldBeFalse()
        (c1 == c5).shouldBeFalse()
        (c1 == c6).shouldBeFalse()
        (c1 == c7).shouldBeFalse()
        (c2 == c3).shouldBeFalse()
        (c2 == c4).shouldBeFalse()
        (c2 == c5).shouldBeFalse()
        (c2 == c6).shouldBeFalse()
        (c2 == c7).shouldBeFalse()
        (c3 == c4).shouldBeFalse()
        (c3 == c5).shouldBeFalse()
        (c3 == c6).shouldBeFalse()
        (c3 == c7).shouldBeFalse()
        (c4 == c5).shouldBeFalse()
        (c4 == c6).shouldBeFalse()
        (c4 == c7).shouldBeFalse()
        (c5 == c6).shouldBeFalse()
        (c5 == c7).shouldBeFalse()
        (c6 == c7).shouldBeFalse()

        c1.hashCode() shouldNotBeEqualTo c2.hashCode()
        c1.hashCode() shouldNotBeEqualTo c3.hashCode()
        c1.hashCode() shouldNotBeEqualTo c4.hashCode()
        c1.hashCode() shouldNotBeEqualTo c5.hashCode()
        c1.hashCode() shouldNotBeEqualTo c6.hashCode()
        c1.hashCode() shouldNotBeEqualTo c7.hashCode()
        c2.hashCode() shouldNotBeEqualTo c3.hashCode()
        c2.hashCode() shouldNotBeEqualTo c4.hashCode()
        c2.hashCode() shouldNotBeEqualTo c5.hashCode()
        c2.hashCode() shouldNotBeEqualTo c6.hashCode()
        c2.hashCode() shouldNotBeEqualTo c7.hashCode()
        c3.hashCode() shouldNotBeEqualTo c4.hashCode()
        c3.hashCode() shouldNotBeEqualTo c5.hashCode()
        c3.hashCode() shouldNotBeEqualTo c6.hashCode()
        c3.hashCode() shouldNotBeEqualTo c7.hashCode()
        c4.hashCode() shouldNotBeEqualTo c5.hashCode()
        c4.hashCode() shouldNotBeEqualTo c6.hashCode()
        c4.hashCode() shouldNotBeEqualTo c7.hashCode()
        c5.hashCode() shouldNotBeEqualTo c6.hashCode()
        c5.hashCode() shouldNotBeEqualTo c7.hashCode()
        c6.hashCode() shouldNotBeEqualTo c7.hashCode()
    }

    @Test
    fun equalingToSpecialCases() {
        val c = newValidEntity()
        (c == c).shouldBeTrue()
    }

    private fun assertInequality(c1: Code, c2: Code) {
        (c1 == c2).shouldBeFalse()
        (c2 == c1).shouldBeFalse()
        c1.hashCode() shouldNotBeEqualTo c2.hashCode()
    }

    private fun assertEquality(c1: Code, c2: Code) {
        (c1 == c2).shouldBeTrue()
        (c2 == c1).shouldBeTrue()
        c1.hashCode() shouldBeEqualTo c2.hashCode()
    }

    @Test
    fun differingValues_withSameComment() {
        val c1 = Code("1A", CODE1, "foo", false, 1, "c1", "", 1)
        val c2 = Code("1A", CODE1, "foo", false, 1, "c1", "", 1)
        assertEquality(c1, c2)
    }

    @Test
    fun differingValues_withDifferingComment() {
        val c1 = Code("1A", CODE1, "foo", false, 1, "c1", "", 1)
        val c2 = Code("1A", CODE1, "bar", false, 1, "c1", "", 1)
        assertInequality(c1, c2)
    }

    @Test
    fun differingValues_withDifferingSort() {
        val c1 = Code("1A", CODE1, "", false, 1, "c1", "", 1)
        val c2 = Code("1A", CODE1, "", false, 1, "c1", "", 2)
        assertInequality(c1, c2)
    }

    companion object {
        private const val CODE1 = "code1"
        private const val CODECLASS10 = "codeclass10"

        private val CREATED = LocalDateTime.parse("2017-01-01T08:01:33.821")
        private val LAST_MOD = LocalDateTime.parse("2017-02-02T08:01:33.821")
    }
}
